codalab/migration.py

```diff
@@ -0,0 +1,216 @@
+
+        self.target_store = self.bundle_manager._model.get_bundle_store(
+            user_id=self.root_user_id, name=self.target_store_name
+        )
+
+        self.target_store_url = self.target_store['url']
+        self.target_store_uuid = self.target_store['uuid']
+
+        assert StorageType.AZURE_BLOB_STORAGE.value == self.target_store['storage_type']
+        self.target_store_type = StorageType.AZURE_BLOB_STORAGE
+
+    def get_bundle_uuids(self, worksheet_uuid):
+        bundle_uuids = self.bundle_manager._model.get_bundle_uuids(
+            {'name': None, 'worksheet_uuid': worksheet_uuid, 'user_id': self.root_user_id},
+            max_results=None,  # return all bundles in the worksheets
+        )
+        return bundle_uuids
+
+    def is_linked_bundle(self, bundle_uuid):
+        bundle_link_url = self.bundle_manager._model.get_bundle_metadata(
+            [bundle_uuid], "link_url"
+        ).get(bundle_uuid)
+        return bundle_link_url is None
+
+    def get_bundle_location(self, bundle_uuid):
+        # bundle_locations = self.bundle_manager._model.get_bundle_locations(bundle_uuid)  # this is getting locations from
+        bundle_location = self.bundle_manager._bundle_store.get_bundle_location(bundle_uuid)
+        return bundle_location
+
+    def get_bundle(self, bundle_uuid):
+        return self.bundle_manager._model.get_bundle(bundle_uuid)
+
+    def get_bundle_info(self, bundle_uuid, bundle_location):
+        target = BundleTarget(bundle_uuid, subpath='')
+        logging.info(f"[migration] {target}")
+        try:
+            info = download_util.get_target_info(bundle_location, target, depth=0)
+            logging.info(f"[migration] {info}")
+        except Exception as e:
+            logging.info(f"[migration] Error: {str(e)}")
+            raise e
+
+        return info
+
+    def upload_to_Azure(self, bundle_uuid, bundle_location, is_dir=False):
+        # generate target bundle path
+        file_name = "contents.tar.gz" if is_dir else "contents.gz"
+        target_location = f"{self.target_store_url}/{bundle_uuid}/{file_name}"
+
+        if FileSystems.exists(target_location):
+            path_util.remove(target_location)
+
+        uploader = BlobStorageUploader(
+            bundle_model=self.bundle_manager._model,
+            bundle_store=self.bundle_manager._bundle_store,
+            destination_bundle_store=None,
+            json_api_client=None,
+        )
+
+        if is_dir:
+            source_fileobj = zip_util.tar_gzip_directory(bundle_location)
+            source_ext = ".tar.gz"
+            unpack = True
+        else:
+            # If it's a file, change it into GzipStream
+            source_fileobj = open(bundle_location, 'rb')
+            source_ext = ''
+            unpack = False
+
+        # Upload file content and generate index file
+        uploader.write_fileobj(source_ext, source_fileobj, target_location, unpack_archive=unpack)
```

Comment:
After my changes to the upload manager, this will modify the disk usage of the user on whose authority this function is called, just FYI. If it's the codalab user, we could just manually reset its disk quota after the migration. I could also add a flag to turn off disk quota checking, but that could potentially be unsafe.

Reply:
One other idea is to decrement the user disk usage in origin_bundle_store, actually. That might even be better.

Reply:
Thanks for reminding me this! I added a function call to decrement the user disk usage when deleting the bundle.